borg/archiver.py

```diff
@@ -607,8 +607,27 @@ def do_break_lock(self, args):
         start to end, or must match from the start to just before a path separator.
         Except for the root path, paths will never end in the path separator when
         matching is attempted.  Thus, if a given pattern ends in a path separator, a
-        '*' is appended before matching is attempted.  Patterns with wildcards should
-        be quoted to protect them from shell expansion.
+        '*' is appended before matching is attempted.
+
+        Regular expressions similar to those found in Perl are also supported. Unlike
+        shell patterns regular expressions are not required to match the complete
+        path and any substring match is sufficient. It is therefore strongly
+        recommended to anchor patterns to the start ('^'), to the end ('$') or both.
+        Path separators ('\\' for Windows and '/' on other systems) in paths always
+        normalized to a forward slash ('/') before applying a pattern. The regular
+        expression syntax is described in the `Python documentation for the re module
+        <https://docs.python.org/3/library/re.html>`_.
+
+        Exclusions can be passed via command line options. The `--exclude` option
+        (`-e` for short) uses shell pattern syntax and `--exclude-regex` uses regular
+        expression syntax. When passed via a shell the patterns should be quoted to
+        protect them from expansion.
+
+        The `--exclude-from` option permits loading exclusion patterns from a text
+        file with one pattern per line. Empty lines as well as lines starting with
+        the number sign ('#') are ignored. If a line starts with '{RE}' (case
```

Comment:
As you write `??:` can be part of a file name (my system has a file named `IO::Socket::SSL::Utils.3pm.gz`, for example). I agree that `{…}` is perhaps not the best prefix even if it's less likely to conflict with file names. I thought about proposing `$`, but then I realized that it doesn't work well when used in combination with shell variables. Please read on.

I'm strongly against requiring the prefix. Other programs, e.g. tar, use shell patterns by default. If they support other syntaxes they usually do so by employing *optional* inline constructs to switch syntax (e.g. `(?m)` for regular expressions in Python) or separate options altogether (as proposed here with `--exclude-regex`). It would be rather unusual for a program to require a prefix in an option *value* to select what could be considered the default glob syntax for Unix systems.

This said I'm in favour of also having an explicit prefix to select fnmatch or glob syntax. This can then be used to circumvent conflicts between a file name and the syntax overloads.

Thus my proposed syntax is as follows:

    [{ "RE" | "FN" | "GL" } ":" ] <pattern>

To preserve compatibility with existing scripts and exclusion files `FN` is the default. The prefix is case-insensitive for ease of use. An actual implementation would interpret all sequences of two letters and a colon at the beginning of the pattern as a prefix and fail if it's not one of the above.

If we do not require the prefix we can also do without `--exclude-regex`. Excluding thumbnails would be `--exclude=re:^/home/[^/]+/\.thumbnails`, old-style fnmatch would be `--exclude=/foo/*/bar` and your improved fnmatch would be accessible via `--exclude=gl:/x/y*.zip`. Should there be a conflict between syntax selection and a file name to be excluded, the syntax has to be explicitly specified: `--exclude=fn:gl:*.zip` (fnmatch syntax, pattern `gl:*.zip`).

The rules would be used for the exclusion file, i.e. `*.zip` for implicit fnmatch, `re:…` and `fn:gl:*.zip`.